File: javascript/test/gilded_rose.test.js
const gildedRose = require("../src/gilded_rose");

test("At the end of each day our system lowers both values for every item", () => {
    const sword = new gildedRose.Item("Sword", 10, 10);
    const shop = new gildedRose.Shop([sword]);
    shop.updateQuality();
    expect(sword.sellIn).toBe(9);
    expect(sword.quality).toBe(9);
}); // coverage: 45%, mutation coverage: 25%

test("Once the sell by date has passed, Quality degrades twice as fast", () => {
    const sword = new gildedRose.Item("Sword", 0, 10);
    const shop = new gildedRose.Shop([sword]);
    shop.updateQuality();
    expect(sword.sellIn).toBe(-1);
    expect(sword.quality).toBe(8);
}); // coverage: 61%, mutation coverage: 39%

test("The Quality of an item is never negative", () => {
    const sword = new gildedRose.Item("Sword", 10, 0);
    const shop = new gildedRose.Shop([sword]);
    shop.updateQuality();
    expect(sword.sellIn).toBe(9);
    expect(sword.quality).toBe(0);
}); // coverage: 61%, mutation coverage: 41%

test("'Aged Brie' actually increases in Quality the older it gets", () => {
    const AgedBrie = new gildedRose.Item("Aged Brie", 10, 10);
    const shop = new gildedRose.Shop([AgedBrie]);
    shop.updateQuality();
    expect(AgedBrie.sellIn).toBe(9);
    expect(AgedBrie.quality).toBe(11);
}); // coverage: 71%, mutation coverage: 51%

test("The Quality of an item is never more than 50", () => {
    const AgedBrie = new gildedRose.Item("Aged Brie", 10, 50);
    const shop = new gildedRose.Shop([AgedBrie]);
    shop.updateQuality();
    expect(AgedBrie.sellIn).toBe(9);
    expect(AgedBrie.quality).toBe(50);
}); // coverage: 71%, mutation coverage: 53%

test("'Sulfuras', being a legendary item, never has to be sold or decreases in Quality", () => {
    const Sulfuras = new gildedRose.Item("Sulfuras, Hand of Ragnaros", 10, 80);
    const shop = new gildedRose.Shop([Sulfuras]);
    shop.updateQuality();
    expect(Sulfuras.sellIn).toBe(10);
    expect(Sulfuras.quality).toBe(80);
}); // coverage: 71%, mutation coverage: 57%

test("'Backstage passes', like aged brie, increases in Quality as its SellIn value approaches", () => {
    const BackstagePasses = new gildedRose.Item("Backstage passes to a TAFKAL80ETC concert", 20, 10);
    const shop = new gildedRose.Shop([BackstagePasses]);
    shop.updateQuality();
    expect(BackstagePasses.sellIn).toBe(19);
    expect(BackstagePasses.quality).toBe(11);
}); // coverage: 77%, mutation coverage: 62%

test("'Backstage passes' quality increases by 2 when there are 10 days or less", () => {
    const BackstagePasses = new gildedRose.Item("Backstage passes to a TAFKAL80ETC concert", 10, 10);
    const shop = new gildedRose.Shop([BackstagePasses]);
    shop.updateQuality();
    expect(BackstagePasses.sellIn).toBe(9);
    expect(BackstagePasses.quality).toBe(12);
}); // coverage: 84%, mutation coverage: 71%

test("'Backstage passes' quality increases by 3 when there are 5 days or less", () => {
    const BackstagePasses = new gildedRose.Item("Backstage passes to a TAFKAL80ETC concert", 5, 10);
    const shop = new gildedRose.Shop([BackstagePasses]);
    shop.updateQuality();
    expect(BackstagePasses.sellIn).toBe(4);
    expect(BackstagePasses.quality).toBe(13);
}); // coverage: 90%, mutation coverage: 76%

test("'Backstage passes' quality drops to 0 after the concert", () => {
    const BackstagePasses = new gildedRose.Item("Backstage passes to a TAFKAL80ETC concert", 0, 10);
    const shop = new gildedRose.Shop([BackstagePasses]);
    shop.updateQuality();
    expect(BackstagePasses.sellIn).toBe(-1);
    expect(BackstagePasses.quality).toBe(0);
}); // coverage: 94%, mutation coverage: 80%

File: java/src/test/java/org/mutationtesting/demo/gildedrose/GildedRoseTest.java
package org.mutationtesting.demo.gildedrose;

import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class GildedRoseTest {
    /**
     * Test spec
     */
    @Test
    public void At_the_end_of_each_day_our_system_lowers_both_values_for_every_item() {
        Item sword = new Item("Sword", 10, 10);
        GildedRose gr = new GildedRose(new Item[]{sword});
        gr.updateQuality();
        assertEquals(9, sword.sellIn);
        assertEquals(9, sword.quality);
    } // coverage: 50%, mutation coverage: 29%

    @Test
    public void Once_the_sell_by_date_has_passed_Quality_degrades_twice_as_fast() {
        Item sword = new Item("Sword", 0, 10);
        GildedRose gr = new GildedRose(new Item[]{sword});
        gr.updateQuality();
        assertEquals(-1, sword.sellIn);
        assertEquals(8, sword.quality);
    } // coverage: 64%, mutation coverage: 42%

    @Test
    public void The_Quality_of_an_item_is_never_negative() {
        Item sword = new Item("Sword", 10, 0);
        GildedRose gr = new GildedRose(new Item[]{sword});
        gr.updateQuality();
        assertEquals(9, sword.sellIn);
        assertEquals(0, sword.quality);
    } // coverage: 64%, mutation coverage: 45%

    @Test
    public void Aged_Brie_actually_increases_in_Quality_the_older_it_gets() {
        Item AgedBrie = new Item("Aged Brie", 10, 10);
        GildedRose gr = new GildedRose(new Item[]{AgedBrie});
        gr.updateQuality();
        assertEquals(9, AgedBrie.sellIn);
        assertEquals(11, AgedBrie.quality);
    } // coverage: 72%, mutation coverage: 53%

    @Test
    public void The_Quality_of_an_item_is_never_more_than_50() {
        Item AgedBrie = new Item("Aged Brie", 10, 50);
        GildedRose gr = new GildedRose(new Item[]{AgedBrie});
        gr.updateQuality();
        assertEquals(9, AgedBrie.sellIn);
        assertEquals(50, AgedBrie.quality);
    } // coverage: 72%, mutation coverage: 55%

    @Test
    public void Sulfuras_being_a_legendary_item_never_has_to_be_sold_or_decreases_in_Quality() {
        Item Sulfuras = new Item("Sulfuras, Hand of Ragnaros", 10, 80);
        GildedRose gr = new GildedRose(new Item[]{Sulfuras});
        gr.updateQuality();
        assertEquals(10, Sulfuras.sellIn);
        assertEquals(80, Sulfuras.quality);
    } // coverage: 72%, mutation coverage: 55%

    @Test
    public void Backstage_passes_like_aged_brie_increases_in_Quality_as_its_SellIn_value_approaches() {
        Item BackstagePasses = new Item("Backstage passes to a TAFKAL80ETC concert", 20, 10);
        GildedRose gr = new GildedRose(new Item[]{BackstagePasses});
        gr.updateQuality();
        assertEquals(19, BackstagePasses.sellIn);
        assertEquals(11, BackstagePasses.quality);
    } // coverage: 78%, mutation coverage: 61%

    @Test
    public void Backstage_passes_quality_increases_by_2_when_there_are_10_days_or_less() {
        Item BackstagePasses = new Item("Backstage passes to a TAFKAL80ETC concert", 10, 10);
        GildedRose gr = new GildedRose(new Item[]{BackstagePasses});
        gr.updateQuality();
        assertEquals(9, BackstagePasses.sellIn);
        assertEquals(12, BackstagePasses.quality);
    } // coverage: 83%, mutation coverage: 66%

    @Test
    public void Backstage_passes_quality_increases_by_3_when_there_are_5_days_or_less() {
        Item BackstagePasses = new Item("Backstage passes to a TAFKAL80ETC concert", 5, 10);
        GildedRose gr = new GildedRose(new Item[]{BackstagePasses});
        gr.updateQuality();
        assertEquals(4, BackstagePasses.sellIn);
        assertEquals(13, BackstagePasses.quality);
    } // coverage: 89%, mutation coverage: 71%

    @Test
    public void Backstage_passes_quality_drops_to_0_after_the_concert() {
        Item BackstagePasses = new Item("Backstage passes to a TAFKAL80ETC concert", 0, 10);
        GildedRose gr = new GildedRose(new Item[]{BackstagePasses});
        gr.updateQuality();
        assertEquals(-1, BackstagePasses.sellIn);
        assertEquals(0, BackstagePasses.quality);
    } // coverage: 92%, mutation coverage: 74%
}
